qiskit/providers/ibmq/job/ibmq_composite_job.py

```diff
@@ -0,0 +1,1274 @@
+        """Retrieve the job result response.
+
+        Args:
+            refresh: If ``True``, re-query the server for the result.
+               Otherwise return the cached value.
+
+        Raises:
+            IBMQJobApiError: If an unexpected error occurred when communicating
+                with the server.
+        """
+        if self._result and not refresh:
+            return self._result
+
+        job_results = [sub_job.result(refresh=refresh, partial=partial)
+                       for sub_job in self._sub_jobs]
+
+        if not partial and any(result is None for result in job_results):
+            return None
+        try:
+            good_result = next(res for res in job_results if res is not None).to_dict()
+        except StopIteration:
+            return None
+
+        ref_expr_result = good_result['results'][0]
+        template_expr_result = {
+            'success': False,
+            'data': {},
+            'status': 'ERROR'
+        }
+        for key in ['shots', 'meas_level', 'seed', 'meas_return']:
+            template_expr_result[key] = ref_expr_result.get(key, None)
+
+        good_result['job_id'] = self.job_id()
+        good_result['success'] = self._status == JobStatus.DONE
+        good_result['results'] = []
+        good_result['status'] = 'PARTIAL COMPLETED' if self._status != JobStatus.DONE \
+            else 'COMPLETED'
+        combined_result = Result.from_dict(good_result)
+
+        for idx, result in enumerate(job_results):
+            if result is not None:
+                combined_result.results.extend(result.results)
+            else:
+                # Get experiment header from Qobj if possible.
+                sub_job = self._sub_jobs[idx]
+                qobj = sub_job.qobj
+                experiments = qobj.experiments if qobj else None
+
+                expr_results = []
+                for circ_idx in range(sub_job.end_index-sub_job.start_index+1):
+                    if experiments:
+                        template_expr_result['header'] = experiments[circ_idx].header.to_dict()
+                    expr_results.append(ExperimentResult.from_dict(template_expr_result))
+                combined_result.results.extend(expr_results)
+
+        return combined_result
+
+    def _get_qobj(self) -> Optional[Union[QasmQobj, PulseQobj]]:
+        """Return the Qobj for this job.
+
+        Returns:
+            The Qobj for this job, or ``None`` if the job does not have a Qobj.
+
+        Raises:
+            IBMQJobApiError: If an unexpected error occurred when retrieving
+                job information from the server.
+        """
+        if not self._qobj:
+            self._qobj = copy.deepcopy(self._sub_jobs[0].qobj)
+            for idx in range(1, len(self._get_circuit_jobs())):
+                self._qobj.experiments.extend(self._sub_jobs[idx].qobj.experiments)
+        return self._qobj
+
+    def _get_circuit_jobs(self) -> List[IBMQCircuitJob]:
```

Comment:
Unfortunately it does need to be recalculated each time because circuit jobs are submitted asynchronously, and `sub_job.job` is set when the submit finishes. So the list of submitted circuit jobs can change each time. I guess I can cache the value after all jobs are submitted, but the list is small so I'm not too worried.